scripts/utils.py

```diff
@@ -337,3 +344,403 @@ def add_population_to_dataframe(
+        Minimum fraction of the total population of the region that each of the listed countries must exceed.
+    min_frac_cumulative_population : float
+        Minimum fraction of the total population of the region that the sum of the listed countries must exceed.
+    countries_regions : pd.DataFrame or None
+        Countries-regions dataset, or None, to load it from owid catalog.
+    population : pd.DataFrame or None
+        Population dataset, or None, to load it from owid catalog.
+
+    Returns
+    -------
+    countries : list
+        List of countries that are expected to have the largest contribution.
+
+    """
+    # TODO: Add tests.
+    if countries_regions is None:
+        countries_regions = catalog.find(
+            "countries_regions", dataset="reference", namespace="owid"
+        ).load()
+
+    if population is None:
+        population = (
+            catalog.find("population", namespace="owid", dataset="key_indicators")
+            .load()
+            .reset_index()
+        )
+
+    # List all countries in the selected region.
+    members = list_countries_in_region(region, countries_regions=countries_regions)
+
+    # Select population data for reference year for all countries in the region.
+    reference = (
+        population[
+            (population["country"].isin(members))
+            & (population["year"] == reference_year)
+        ]
+        .dropna(subset="population")
+        .sort_values("population", ascending=False)
+        .reset_index(drop=True)
+    )
+
+    # Calculate total population in the region, and the fractional contribution of each country.
+    total_population = reference["population"].sum()
+    reference["fraction"] = reference["population"] / total_population
+
+    # Select countries that exceed a minimum individual fraction of the total population of the region.
+    selected = reference[
+        (reference["fraction"] > min_frac_individual_population)
+    ].reset_index(drop=True)
+
+    # Among remaining countries, select countries that, combined, exceed a minimum fraction of the total population.
+    selected["cumulative_fraction"] = selected["population"].cumsum() / total_population
+    candidates_to_ignore = selected[
+        selected["cumulative_fraction"] > min_frac_cumulative_population
+    ]
+    if len(candidates_to_ignore) == 0:
+        print(
+            "WARNING: Conditions are too strict to select countries that must be included in the data."
+        )
+    else:
+        # Select the smallest possible list of countries that fulfil both conditions.
+        selected = selected.loc[0 : candidates_to_ignore.index[0]]
+
+    print(
+        f"{len(selected)} countries must be informed for {region} (covering {selected['fraction'].sum() * 100: .2f}% "
+        f"of the population; otherwise aggregate data will be nan."
+    )
+    countries = selected["country"].tolist()
+
+    return countries
+
+
+def add_region_aggregates(
```

Comment:
Would it make sense to split this into two functions - one that returns `countries_that_must_be_present` and a second one that does the aggregation? It would take two lines instead of one, but also be more flexible.

For instance when doing exploration in a notebook I'd like to print those countries to see what's going on. Or maybe define my own list of countries or use `countries_that_must_be_present=None` when I don't care about missing regions and just want to do aggregation.

Reply:
Good idea, I'll see how that would look.

Reply:
Done.